### PR TITLE
fix(java): make license header fixing idempotent

### LINKED ISSUE #1007: Java: licence header fixing should be idempotent
Synthtool tries to fix grpc and proto license headers. As we migrate fixes into the generator, we should handle when the generator starts providing the headers and not duplicate the license headers.

### PR DESCRIPTION
**Pull request by chingor13**
Fixes #1007

### CODE REVIEW

File: synthtool/languages/java.py [hunk 1]
```diff
@@ -92,9 +91,30 @@ def _download_formatter(version: str, dest: Path) -> None:
...
+
+def _file_has_header(path: Path) -> bool:
+    with open(path, "rt") as fp:
+        for line in fp:
+            if HEADER_REGEX.search(line):
+                return True
+    return False
+
+
+# return a subset of files that do not already have a header
```
Comment (SurferJeffAtGoogle) [inline synthtool/languages/java.py:105]: Move this comment to be a doc string on line 107.

File: synthtool/languages/java.py [hunk 2]
```diff
@@ -106,9 +126,13 @@ def fix_proto_headers(proto_root: Path) -> None:
     )
 
 
-def fix_grpc_headers(grpc_root: Path, package_name: str) -> None:
+def fix_grpc_headers(grpc_root: Path, package_name: str = "unused") -> None:
```
Comment (SurferJeffAtGoogle) [inline synthtool/languages/java.py:129]: Please add a doc comment string for this function.

File: synthtool/languages/java.py [hunk 3]
```diff
@@ -23,12 +23,11 @@
...
 from pathlib import Path
-from typing import Any, Optional, Dict, List
+from typing import Any, Optional, Dict, Iterable, List
 
 JAR_DOWNLOAD_URL = "https://github.com/google/google-java-format/releases/download/google-java-format-{version}/google-java-format-{version}-all-deps.jar"
 DEFAULT_FORMAT_VERSION = "1.7"
-GOOD_LICENSE = """
-/*
+GOOD_LICENSE = """/*
  * Copyright 2020 Google LLC
```
Comment (SurferJeffAtGoogle) [inline synthtool/languages/java.py:31]: Will we have to update "2020" at some future date?

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: synthtool/languages/java.py (hunk 3 above)
Comment (chingor13) [reply synthtool/languages/java.py:31]: Yes, but wanted to keep it separate from this bugfix

Comment (SurferJeffAtGoogle) [reply synthtool/languages/java.py:31]: Ok.

File: synthtool/languages/java.py (hunk 1 above)
Comment (chingor13) [reply synthtool/languages/java.py]: Done

File: synthtool/languages/java.py (hunk 2 above)
Comment (chingor13) [reply synthtool/languages/java.py:131]: Done

File: synthtool/languages/java.py [hunk 4]
```diff
@@ -92,9 +91,32 @@ def _download_formatter(version: str, dest: Path) -> None:
...
+    for path in paths:
+        anchor = Path(path.anchor)
+        remainder = str(path.relative_to(path.anchor))
+        for file in anchor.glob(remainder):
+            if not _file_has_header(file):
+                yield file
+
+
 def fix_proto_headers(proto_root: Path) -> None:
+    """Helper to ensure that generated proto classes have appropriate license headers."""
```
Comment (SurferJeffAtGoogle) [inline synthtool/languages/java.py:117]: In the doc string, please describe what the function does.  What actions does it take to fix the proto headers?

File: synthtool/languages/java.py [hunk 5]
```diff
@@ -106,9 +128,12 @@ def fix_proto_headers(proto_root: Path) -> None:
     )
 
 
-def fix_grpc_headers(grpc_root: Path, package_name: str) -> None:
+def fix_grpc_headers(grpc_root: Path, package_name: str = "unused") -> None:
+    """Helper to ensure that generated grpc stub classes have appropriate license headers."""
```
Comment (SurferJeffAtGoogle) [inline synthtool/languages/java.py:132]: ditto

File: synthtool/languages/java.py (hunk 4 above)
Comment (chingor13) [reply synthtool/languages/java.py]: Done

File: synthtool/languages/java.py (hunk 5 above)
Comment (chingor13) [reply synthtool/languages/java.py]: Done.

**Review by bcoe — APPROVED** [review]

File: synthtool/languages/java.py [hunk 6]
```diff
@@ -23,12 +23,11 @@
...
 from synthtool.log import logger
 from pathlib import Path
-from typing import Any, Optional, Dict, List
+from typing import Any, Optional, Dict, Iterable, List
 
 JAR_DOWNLOAD_URL = "https://github.com/google/google-java-format/releases/download/google-java-format-{version}/google-java-format-{version}-all-deps.jar"
 DEFAULT_FORMAT_VERSION = "1.7"
-GOOD_LICENSE = """
-/*
+GOOD_LICENSE = """/*
```
Comment (bcoe) [inline synthtool/languages/java.py:30]: I notice that GOOD_LICENSE has 2020 hardcoded into it, do we need to update this date over time? how does this work for new libraries.

**Review by SurferJeffAtGoogle — APPROVED** [review]

**Pull request merged**